### PR TITLE
[RFR]: Getting the SHA to the deb and rpm packages

### PR DESCRIPTION
**Pull request by humblearner**
- Script for creating package.meta in st2

### CODE REVIEW

**Comment by Kami** [comment]
Nice :+1: 

I'm definitely a big proponent of that, because as I mentioned before, every build should be reproducible. This means we should be able to track package to a specific git revision and we should be able to rebuild packages (and get exactly the same bits) if we re-build packages for a specific revision.

**Comment by Kami** [comment]
Is there a corresponding PR in st2-packages or that's something which still needs to be done?

I believe we also need to be a bit careful when we run this script and read out st2 component version, since (at least in the past) this version was written as part of the build which means we would need to run this script after that step.

**Comment by humblearner** [comment]
@Kami : Thanks to @lakshmi and @manasdk for proposing this idea and helping me with it.

> Is there a corresponding PR in st2-packages or that's something which still needs to be done?

Yes, Lakshmi suggested here: https://github.com/StackStorm/st2-packages/blob/master/rake/build/package_st2.rake#L34 

I will add this line here:
` 35               execute :bash, '$GITDIR/populate-sha.sh' `

**Comment by armab** [comment]
:+1:

Related to this issue: https://github.com/StackStorm/st2-packages/issues/148

File: scripts/populate-sha.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+git_sha=$(git rev-parse --short HEAD)
+version=$(cat ../st2common/st2common/__init__.py | grep __version__ | sed -r '1!d;s/.*=\s+//'| tr -d "'")
```
Comment (armab) [inline scripts/populate-sha.sh]:
```py
python -c 'execfile("../st2common/st2common/__init__.py"); print __version__'
```
looks more clear.

See: https://github.com/StackStorm/st2-packages/search?utf8=%E2%9C%93&q=print+__version__

**Comment by humblearner** [comment]
Sorry for all this mess. I am unable to ssh into the CircleCI builds. Hence, trying this way. So far got here: 
https://circleci.com/gh/StackStorm/st2-packages/880

```
Starting Packages Build for wheezy ...
+ docker-compose -f docker-compose.circle.yml run -e ST2_GITURL=https://github.com/StackStorm/st2 -e ST2_GITREV=package/sha_stuff -e ST2PKG_VERSION=1.5dev -e ST2PKG_RELEASE=53 -e ST2MISTRAL_GITURL=https://github.com/StackStorm/mistral -e ST2MISTRAL_GITREV=master -e MISTRAL_VERSION=1.5dev -e MISTRAL_RELEASE=53 -e RABBITMQHOST=172.17.0.1 -e POSTGRESHOST=172.17.0.1 -e MONGODBHOST=172.17.0.1 -e 'ST2_PACKAGES=st2 st2mistral' -e ST2_CIRCLE_URL=https://circleci.com/gh/StackStorm/st2-packages/880 wheezy build
!
!
[checkout: update st2]      Writing version, SHA and circle_build_url  to ../package.meta  ...
[checkout: update st2]      Contents of package.meta:
[checkout: update st2]      [server]
[checkout: update st2]      version = 1.5dev
[checkout: update st2]      git_sha = 5d21d14
[checkout: update st2]      circle_build_url=
[checkout: update st2]      ~
[checkout: update st2]      env variables to find Circle CI URL
[checkout: update st2]      declare -x ARTIFACT_DIR="/root/build"
[checkout: update st2]      declare -x BASEDIR="/root"
[checkout: update st2]      declare -x DEBUG_LEVEL="0"
[checkout: update st2]      declare -x GITDIR="/tmp/st2-20160520-11-1q4q5ga"
[checkout: update st2]      declare -x GITREV="package/sha_stuff"
[checkout: update st2]      declare -x GITURL="https://github.com/StackStorm/st2"
[checkout: update st2]      declare -x HOME="/root"
[checkout: update st2]      declare -x LOGNAME="root"
[checkout: update st2]      declare -x MAIL="/var/mail/root"
[checkout: update st2]      declare -x MISTRAL_RELEASE="53"
[checkout: update st2]      declare -x MISTRAL_VERSION="1.5dev"
[checkout: update st2]      declare -x MONGODBHOST="172.17.0.1"
[checkout: update st2]      declare -x OLDPWD="/tmp/st2-20160520-11-1q4q5ga/scripts"
[checkout: update st2]      declare -x PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin/X11"
[checkout: update st2]      declare -x POSTGRESHOST="172.17.0.1"
[checkout: update st2]      declare -x PWD="/root"
[checkout: update st2]      declare -x RABBITMQHOST="172.17.0.1"
[checkout: update st2]      declare -x SHELL="/bin/bash"
[checkout: update st2]      declare -x SHLVL="2"
[checkout: update st2]      declare -x SSH_AUTH_SOCK="/tmp/ssh-pNTovyCrtU/agent.6"
[checkout: update st2]      declare -x SSH_CLIENT="172.17.0.4 45074 22"
[checkout: update st2]      declare -x SSH_CONNECTION="172.17.0.4 45074 172.17.0.3 22"
[checkout: update st2]      declare -x ST2PKG_RELEASE="53"
[checkout: update st2]      declare -x ST2PKG_VERSION="1.5dev"
[checkout: update st2]      declare -x ST2_PYTHON="0"
[checkout: update st2]      declare -x USER="root"
[checkout: update st2]      declare -x WHEELDIR="/tmp/wheelhouse"
```
Now just need **ST2_CIRCLE_URL=https://circleci.com/gh/StackStorm/st2-packages/880** to show up in package.meta : _circle_build_url = ????_

**Comment by armab** [comment]
To SSH into builds you need to have `admin` permissions for github `st2-packages` repo, also don't forget to log in to CircleCI via Github.
After that you'll see `Rebuild with SSH button` in top right corner: https://i.imgur.com/crXjCUE.png

Hope that helps.

**Comment by Kami** [comment]
@armab Thanks for clarifying.

@humblearner I just checked and you are part of the core team which has admin access to this repo so it's probably related to you not being logged to CircleCI via Github or similar.

**Comment by armab** [comment]
Answered in: 
https://github.com/StackStorm/st2-packages/pull/299

**Comment by humblearner** [comment]
@Kami @armab : Thanks!! I tried all that yesterday, but issue was from CircleCI's side, ssh was not going through or maybe the corporate network:

```
lkannan@HQ2-LC02QV2NDFVH8 ~ $ ssh -vvv -i ~/.ssh/id_rsa -p 64576 ubuntu@54.158.38.191                         [ruby-2.0.0p648]
OpenSSH_6.9p1, LibreSSL 2.1.8
debug1: Reading configuration data /Users/lkannan/.ssh/config
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug2: ssh_connect: needpriv 0
debug1: Connecting to 54.158.38.191 [54.158.38.191] port 64576.
debug1: Connection established.
debug1: identity file /Users/lkannan/.ssh/id_rsa type 1
debug1: key_load_public: No such file or directory
debug1: identity file /Users/lkannan/.ssh/id_rsa-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_6.9
^[[C


ssh_exchange_identification: read: Operation timed out
lkannan@HQ2-LC02QV2NDFVH8 ~ $
```

File: scripts/populate-sha.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+git_sha = $(git rev-parse --short HEAD)
+circle_build_url = $(echo $ST2_CIRCLE_URL)
+
+package_meta
+
+echo Contents of package.meta:
+cat ../package.meta
+popd
+
+echo env variables to find Circle CI URL
```
Comment (lakshmi-kannan) [inline scripts/populate-sha.sh]: Please remove this to reduce noise.

File: st2common/packaging/debian/install [hunk 3]
```diff
@@ -18,3 +18,4 @@ tools/migrate_triggers_to_include_ref_count.py usr/lib/python2.7/dist-packages/s
 tools/migrate_messaging_setup.py usr/lib/python2.7/dist-packages/st2common/bin/
 rbac/roles/sample.yaml opt/stackstorm/rbac/roles/
 rbac/assignments/sample.yaml opt/stackstorm/rbac/assignments/
+package.meta opt/stackstorm/st2/
```
Comment (lakshmi-kannan) [inline st2common/packaging/debian/install]: Remove

File: st2common/packaging/rpm/st2common-rhel6.spec [hunk 4]
```diff
@@ -52,6 +52,7 @@ cp -R contrib/examples %{buildroot}/usr/share/doc/st2/
 chmod -R 775 %{buildroot}/usr/share/doc/st2/examples
 cp -R st2common %{buildroot}/%{python2_sitelib}/
 cp -R bin %{buildroot}/%{python2_sitelib}/st2common/
+cp package.meta %{buildroot}/opt/stackstorm/st2/
```
Comment (lakshmi-kannan) [inline st2common/packaging/rpm/st2common-rhel6.spec]: Remove

File: scripts/populate-package-meta.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+
+pushd $GITDIR/scripts
+
+config=../package.meta
+echo "Writing version, gitSHA and circle_build_url to: package.meta (Final location: /opt/stackstorm/st2)"
```
Comment (lakshmi-kannan) [inline scripts/populate-package-meta.sh]: /opt/stackstorm/st2/package.meta

Comment (lakshmi-kannan) [inline scripts/populate-package-meta.sh]: Remove that final location thing completely.

**Comment by lakshmi-kannan** [comment]
👍  LGTM

**Pull request merged**